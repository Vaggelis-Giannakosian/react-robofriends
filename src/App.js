import React from "react";
import CardList from "./CardList/CardList";
import SearchBox from "./SearchBox/SearchBox";
import './App.css'

class App extends React.Component {

    constructor() {
        super();
        this.state = {
            robots: [],
            searchField: ''
        }
    }

    componentDidMount = async () => {
        fetch('https://jsonplaceholder.typicode.com/users')
            .then(response => response.json())
            .then(users => this.setState({ robots: users }));
    }


    onSearchChange = ({target: {value}}) => {
        this.setState({
            searchField: value
        });
    }

    render() {

        if( !  this.state.robots.length){
            return (<h1>Loading...</h1>);
        }

        const filteredRobots = this.state.robots.filter(robot => {
            return robot.name.toLowerCase().includes(this.state.searchField.toLowerCase())
        })

        return (
            <div className="tc">
                <h1 className="f1">RoboFriends</h1>
                <SearchBox searchChange={this.onSearchChange}/>
                <CardList robots={filteredRobots}/>
            </div>
        );
    }

}


export default App