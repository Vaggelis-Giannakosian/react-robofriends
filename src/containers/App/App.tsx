// import React from "react";
import React, {useState, useEffect, ChangeEvent} from 'react';
import {useSelector, useDispatch} from 'react-redux';
// import {connect} from 'react-redux'
import {setSearchField, requestRobots} from "../../actions";
import MainPage from "../../components/MainPage/MainPage";
import {Dispatch} from "redux";
import {AppState} from "../../index";
// import Card from "../../components/card/Card";

/**
 * App Container With local state - function syntax
 * @param props
 */

// const App =  (props)  => {
//
//     const [robots, setRobots] = useState([])
//     const [searchField, setSearchField] = useState('')
//
//     useEffect(() => {
//         fetch('https://jsonplaceholder.typicode.com/users')
//             .then(response => response.json())
//             .then(users => setRobots(users));
//     }, [])
//
//     function onSearchChange({target: {value}}) {
//         setSearchField(value)
//     }
//
//     if (!robots.length) {
//         return (<h1>Loading...</h1>);
//     }
//
//     const filterRobots = robots.filter(robot => {
//         return robot.name.toLowerCase().includes(searchField.toLowerCase())
//     })
//
//     return (
//      <MainPage searchChange={onSearchChange} robots={ filterRobots } />
//     );
// }


/**
 * App container with redux store - function syntax
 * @param state
 */

export interface IRobot {
    id: number,
    name: string,
    email: string
}

const App = () => {

    const robosUsers: IRobot[] = useSelector((state: AppState) => state.requestRobots.robots)
    const isPending = useSelector((state: AppState) => state.requestRobots.isPending)
    const text = useSelector((state: AppState) => state.searchRobots.searchField)
    const dispatch : Dispatch<any>  = useDispatch();
    const [searchResults, setSearchResults] = useState<IRobot[]>([]);



    const onSearchChange = (e: ChangeEvent<HTMLInputElement>) :void  => {
        dispatch(setSearchField(e.target.value))
    };

    useEffect(() => {
        dispatch(requestRobots());
    }, [dispatch])

    useEffect(() => {

        const filteredRobots: IRobot[] = robosUsers.length ?
            robosUsers.filter((robot: IRobot) => robot.name.toLowerCase().includes(text.toLowerCase())) :
            [];

        setSearchResults(filteredRobots);
    }, [text, robosUsers])

    if (isPending || !searchResults) {
        return (<h1>Loading...</h1>);
    }

    return (
        <MainPage searchChange={onSearchChange} robots={searchResults}/>
    );
}
export default App;


/**
 * App container With redux store - class syntax
 * @param state
 */
// const mapStateToProps = state => ({
//     searchField: state.searchRobots.searchField,
//     robots: state.requestRobots.robots,
//     isPending: state.requestRobots.isPending,
//     error: state.requestRobots.error,
// })
//
// const mapDispatchToProps = dispatch => {
//     return {
//         onSearchChange: event => dispatch(setSearchField(event.target.value)),
//         onRequestRobots: () => dispatch(requestRobots())
//     }
// }
//
// class App extends React.Component {
//
//
//     componentWillMount = async () => {
//         // fetch('https://jsonplaceholder.typicode.com/users')
//         //     .then(response => response.json())
//         //     .then(users => this.setState({robots: users}));
//         this.props.onRequestRobots()
//     }
//
//     // onSearchChange = ({target: {value}}) => {
//     //     this.setState({
//     //         searchField: value
//     //     });
//     // }

//     filterRobots(robots, searchField) {
//         return robots.filter(robot => {
//             return robot.name.toLowerCase().includes(searchField.toLowerCase())
//         })
//     }
//
//     render() {
//
//         const {searchField,robots, isPending, onSearchChange} = this.props;
//
//         if ( isPending || !robots.length) {
//             return (<h1>Loading...</h1>);
//         }
//
//         const filteredRobots = this.filterRobots(robots, searchField);
//
//         return (
//              <MainPage searchChange={onSearchChange} robots={ filteredRobots } />
//         );
//
//     }
//
// }

// export default connect(mapStateToProps, mapDispatchToProps)(App);
