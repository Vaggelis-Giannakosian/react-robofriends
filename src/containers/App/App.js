import React from "react";
// import React, {useState, useEffect} from "react";
import {connect} from 'react-redux'
import CardList from "../../components/CardList/CardList";
import SearchBox from "../../components/SearchBox/SearchBox";
import Scroll from '../../components/Scroll/Scroll'
import ErrorBoundry from '../../components/ErrorBoundry'

import './App.css'
import {setSearchField} from "../../actions";


// const App = (props) => {
//
//     const [robots, setRobots] = useState([])
//     const [searchField, setSearchField] = useState('')
//
//     useEffect(() => {
//         fetch('https://jsonplaceholder.typicode.com/users')
//             .then(response => response.json())
//             .then(users => setRobots(users));
//     }, [])
//
//     function onSearchChange({target: {value}}) {
//         setSearchField(value)
//     }
//
//     if (!robots.length) {
//         return (<h1>Loading...</h1>);
//     }
//
//     const filterRobots = robots.filter(robot => {
//         return robot.name.toLowerCase().includes(searchField.toLowerCase())
//     })
//
//     return (
//         <div className="tc">
//             <h1 className="f1">RoboFriends</h1>
//             <SearchBox searchChange={onSearchChange}/>
//             <Scroll>
//                 <ErrorBoundry>
//                     <CardList robots={filterRobots}/>
//                 </ErrorBoundry>
//             </Scroll>
//         </div>
//     );
// }

const mapStateToProps = state => ({
    searchField: state.searchField
})

const mapDispatchToProps = dispatch => ({
    onSearchChange: event => dispatch(setSearchField(event.target.value))
})

class App extends React.Component {

    constructor() {
        super();
        this.state = {
            robots: []
        }
    }

    componentDidMount = async () => {
        fetch('https://jsonplaceholder.typicode.com/users')
            .then(response => response.json())
            .then(users => this.setState({robots: users}));
    }

    // onSearchChange = ({target: {value}}) => {
    //     this.setState({
    //         searchField: value
    //     });
    // }

    filterRobots(robots, searchField) {
        return robots.filter(robot => {
            return robot.name.toLowerCase().includes(searchField.toLowerCase())
        })
    }

    render() {

        const {robots} = this.state
        const {searchField, onSearchChange} = this.props;

        if (!robots.length) {
            return (<h1>Loading...</h1>);
        }

        const filteredRobots = this.filterRobots(robots, searchField);

        return (
            <div className="tc">
                <h1 className="f1">RoboFriends</h1>
                <SearchBox searchChange={onSearchChange}/>
                <Scroll>
                    <ErrorBoundry>
                        <CardList robots={filteredRobots}/>
                    </ErrorBoundry>
                </Scroll>
            </div>
        );

    }

}


export default connect(mapStateToProps, mapDispatchToProps)(App);