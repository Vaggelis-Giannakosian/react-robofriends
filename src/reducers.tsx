import {
    CHANGE_SEARCH_FIELD,
    REQUEST_ROBOTS_SUCCESS,
    REQUEST_ROBOTS_FAILED, REQUEST_ROBOTS_PENDING
} from './constants'
import {IRobot} from "./containers/App/App";


interface ISearchState {
    searchField: string
}

const initialStateSearch: ISearchState = {
    searchField: '',
}

interface ISearchAction {
    type?: string,
    payload?: string
}

export const searchRobots = (state: ISearchState = initialStateSearch, action: ISearchAction = {}) => {
    switch (action.type) {
        case CHANGE_SEARCH_FIELD:
            return Object.assign({}, state, {searchField: action.payload})
        default:
            return state;
    }
}

interface IRequestState {
    isPending: boolean,
    robots: IRobot[],
    error: string
}

const initialStateRobots: IRequestState = {
    isPending: false,
    robots: [],
    error: ''
}

interface IRequestAction {
    type?: string,
    payload?: string
}

export const requestRobots = (state: IRequestState = initialStateRobots, action: IRequestAction = {}) => {
    switch (action.type) {
        case REQUEST_ROBOTS_PENDING:
            return Object.assign({}, state, {isPending: true})
        case REQUEST_ROBOTS_SUCCESS:
            return Object.assign({}, state, {isPending: false, robots: action.payload})
        case REQUEST_ROBOTS_FAILED:
            return Object.assign({}, state, {isPending: false, error: action.payload})
        default:
            return state;
    }
}
