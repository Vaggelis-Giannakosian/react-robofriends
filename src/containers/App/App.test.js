import App from './App'
import {shallow, mount, render} from 'enzyme'
import {Provider} from "react-redux";
import {applyMiddleware, combineReducers, createStore} from "redux";
import thunkMiddleware from "redux-thunk";
import {requestRobots, searchRobots} from "../../reducers";


const rootReducer = combineReducers({searchRobots,requestRobots})
const store = createStore(rootReducer,  applyMiddleware(thunkMiddleware))

it('App',()=>{
    expect(mount( <Provider store={store}>
                  <App/>
              </Provider>).length).toEqual(1)


    expect(mount( <Provider store={store}>
        <App/>
    </Provider>)).toMatchSnapshot()
})