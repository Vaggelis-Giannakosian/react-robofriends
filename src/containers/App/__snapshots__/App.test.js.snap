// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <App>
    <h1>
      Loading...
    </h1>
  </App>
</Provider>
`;
